### DISCUSSION TITLE
[Bug] v4 (GraphQL) API: Trouble retrieving pull request review comments

### DISCUSSION THREAD

**Discussion by nporteschaikin** [community/community#24666] [Apps, API and Webhooks]
<p>Howdy! I’m running into a couple of issues using your (wonderful) v4 API. I <a href="https://platform.github.community/t/pullrequestreviewthread-no-data/6170" rel="nofollow noopener">found</a> a <a href="https://platform.github.community/t/pullrequestreviewthread-no-longer-showing-up-in-timeline-for-prs/2755" rel="nofollow noopener">couple</a> of possibly-related threads on your read-only platform forum, but I wanted to mention them here in hopes that there is an available solution.</p>
<p>The crux of what I’m trying to do is I’m trying to retrieve all <code>PullRequestReviewComment</code> objects associated with a <code>PullRequestReview</code>. I’ve tried to achieve this in three separate ways:</p>
<ul>
<li>
<p>By querying <code>PullRequestReviewComment</code> objects returned as part of the pull request’s timeline:</p>
<p>query($owner: String!, $name: String!, $number: Int!) {<br>
repository(owner: $owner, name: $name) {<br>
pullRequest: issueOrPullRequest(number: $number) {<br>
… on PullRequest {<br>
timeline(last: 100) {<br>
… on PullRequestReviewComment {<br>
id<br>
createdAt<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}</p>
</li>
</ul>
<p>Surprisingly, this always returns <strong>zero</strong> <code>PullRequestReviewComment</code> objects.</p>
<ul>
<li>
<p>By querying <code>PullRequestReviewComment</code> objects returned as part of each retrieved <strong>pull request review</strong> in the pull request’s timeline:</p>
<p>query($owner: String!, $name: String!, $number: Int!) {<br>
repository(owner: $owner, name: $name) {<br>
pullRequest: issueOrPullRequest(number: $number) {<br>
… on PullRequest {<br>
timeline(last: 100) {<br>
… on PullRequestReview {<br>
comments(first: 100) {<br>
nodes {<br>
id<br>
createdAt<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}</p>
</li>
</ul>
<p>Even more strangely, this returns  <strong>some </strong> of the pull request comments. The response seems to be omitting all but the first comment associated with each commented-on line.</p>
<ul>
<li>
<p>By querying <code>PullRequestReviewComment</code> objects returned as part of each retrieved <strong>pull request thread</strong> in the pull request’s timeline:</p>
<p>query($owner: String!, $name: String!, $number: Int!) {<br>
repository(owner: $owner, name: $name) {<br>
pullRequest: issueOrPullRequest(number: $number) {<br>
… on PullRequest {<br>
timeline(last: 100) {<br>
… on PullRequestReviewThread {<br>
comments(first: 100) {<br>
nodes {<br>
id<br>
createdAt<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}<br>
}</p>
</li>
</ul>
<p>This solution made a bit of sense given the second solution’s incompleteness: comments are likely threaded, given the <code>replyTo</code> field in the <code>PullRequestReviewComment</code> definition. Alas, this always returned  <strong>zero </strong> threads and thus zero comments.</p>
<p>I’ve considered <a href="https://github.com/GitHawkApp/GitHawk/issues/337#issuecomment-339491871" rel="nofollow noopener">one possible workaround</a>, which is to use the v3 API (as it apparently returns <em>all</em> review comments related to a pull request), but we are consuming enough data per token from your API that this wouldn’t perform at scale. Hoping to find a solution that allows us to continue to query the v4 API for this data.</p>

**Comment by nickvanw** [comment]
<p>Hey <a class="mention" href="/nporteschaikin">@nporteschaikin</a>!</p>
<p>I am glad you are finding the v4 API wonderful - I think you’ve found a less-than-wonderful gap in the information that’s available. Looking at it today, I think the way to get this information would be to do a lot of work using the <code>reviews</code> connection with a query like this:</p>
<pre><code>query ($owner: String!, $name: String!, $number: Int!) {
  repository(owner: $owner, name: $name) {
    pullRequest: issueOrPullRequest(number: $number) {
      ... on PullRequest {
        reviews(last: 100) {
          totalCount
          nodes {
            id
            author {
              login
            }
            url
            comments(first: 100) {
              nodes {
                id
                replyTo {
                  id
                }
                body
                author {
                  login
                }
              }
            }
          }
        }
      }
    }
  }
} 
</code></pre>
<p>If you look at the id of the `eplyTo field on the comments, you should be able to take a look at what comments and reviews are “top level”, and what comments are attributed to them.</p>
<p>Unfortunately, this is not a great UX! As you’ve discovered in those earlier issues, this has been an issue for quite some time (sorry <a class="mention" href="/rnystrom">@rnystrom</a> and <a class="mention" href="/yakov116">@yakov116</a>, <a class="mention" href="/hsborges">@hsborges</a> and everyone else who we let down this time and others).</p>
<p>I’ve got some GraphQL Schema in development that would allow someone to query something like this:</p>
<pre><code>query ($owner: String!, $name: String!, $number: Int!) {
  repository(owner: $owner, name: $name) {
    pullRequest: issueOrPullRequest(number: $number) {
      __typename
      ... on PullRequest {
        reviewThreads(first: 100) {
          edges {
            node {
              __typename
              comments(first: 10) {
                nodes {
                  pullRequestReview {
                    id
                  }
                  body
                }
              }
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p>Which, for the infamous issue <a href="https://github.com/GitHawkApp/GitHawk/pull/555" rel="nofollow noopener">https://github.com/GitHawkApp/GitHawk/pull/555</a> as mentioned in other threads, would return something like this:</p>
<pre><code>{
  "data": {
    "repository": {
      "pullRequest": {
        "__typename": "PullRequest",
        "reviewThreads": {
          "edges": [
            {
              "node": {
                "__typename": "PullRequestReviewThread",
                "comments": {
                  "nodes": [
                    {
                      "pullRequestReview": {
                        "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3NjkyNTI5NzE="
                      },
                      "body": "Can we remove the double space, and if it fails it should default to just returning `fileName`"
                    },
                    {
                      "pullRequestReview": {
                        "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3NjkyNTkxMTk="
                      },
                      "body": "I have tidied up the code and removed the `potentialLanguage`. I was planning to use that for the 'you know what' PR as Highlight.js is not correctly automatically detected the Swift language 😅\r\n\r\nAt GitHub they are using their own Ruby library [github/linguist](https://github.com/github/linguist) can't ask the Github API to send the language?"
                    },
                    {
                      "pullRequestReview": {
                        "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3NjkzOTkwNTE="
                      },
                      "body": "Leaving this reply as a test for PR review threads, demonstrating how broken they are 😞 "
                    }
                  ]
                }
              }
            },
            {
              "node": {
                "__typename": "PullRequestReviewThread",
                "comments": {
                  "nodes": [
                    {
                      "pullRequestReview": {
                        "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3NjkyNTMwMTc="
                      },
                      "body": "Probably want to extract a list of known languages to an enum + separate file? And maybe we can already find a list like that somewhere?\r\n\r\nThen we can `return Language(rawValue: fileExtension)` (or something similar, eg. when there are multiple extensions for one language) instead of this weird `switch`. :D"
                    }
                  ]
                }
              }
            },
            {
              "node": {
                "__typename": "PullRequestReviewThread",
                "comments": {
                  "nodes": [
                    {
                      "pullRequestReview": {
                        "id": "MDE3OlB1bGxSZXF1ZXN0UmV2aWV3NjkyNTI5NzE="
                      },
                      "body": "Not sure this is necessary, at least yet? ~Also it shouldn't **only** support Swift, even if we do keep it~\r\n\r\nEdit// See BasThomas comment"
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  }
}
</code></pre>
<p> How does that sound? I’ll keep this thread up to date as we look to release this.</p>
<p>Separately, to people I’ve mentioned in this issue (including you <a class="mention" href="/nporteschaikin">@nporteschaikin</a>): I am sure you’ve got more schema suggestions, frustrations, or anything of the sort. If you’re interested in chatting about them with me, I’d love to sit down over a video chat and go through what’s got you down with our public GraphQL API.</p>

**Comment by yakov116** [comment]
<p><a class="mention" href="/k0shk0sh">@k0shk0sh</a> do you have any input for us?</p>

**Comment by nporteschaikin** [comment]
<p>Hi <a class="mention" href="/nickvanw">@nickvanw</a>,</p>
<p>Thanks for getting back to me and for your detailed response! We’re heavy users of your GraphQL API at Code Climate and it’s been mostly a pleasant experience :slight_smile:</p>
<blockquote>
<p><a class="mention" href="/nickvanw">@nickvanw</a> wrote:</p>
</blockquote>
<blockquote>
<p>If you look at the id of the `eplyTo field on the comments, you should be able to take a look at what comments and reviews are “top level”, and what comments are attributed to them.</p>
</blockquote>
<p>In my experience, this wouldn’t solve the issue we’re running into. It looks like the response we receive from GraphQL (mostly) does not include <em>replies</em> to comments - rather, we’re receiving all top-level comments, which we can’t connect to their replies. Please correct me if I’m misunderstanding something! A <code>replies</code> connection would be another (inelegant) option, but so would the <code>reviewThreads</code> connection you proposed below. </p>
<p>For now, we’re using the v3 API, although - as you could imagine - it’s significantly more challenging to extract this data at scale via the traditional REST API. I look forward to seeing the <code>reviewThreads</code> connection released!</p>
<p>As for additional feedback on the GraphQL API, I’d love to hop on a call and chat! As alluded to above, we’re extracting a lot of data, regularly, from the v4 API and I think we’ve collected a lot of wisdom, best practices, and feedback for GitHub. I’d love to shoot you an email - let me know how to reach you privately!</p>

**Comment by nickvanw** [comment]
<p>Hey <a class="mention" href="/nporteschaikin">@nporteschaikin</a>,</p>
<p>I’m happy to hear that! Perhaps we may be speaking about different things - looking at the query I posted earlier that works now, I believe one could re-create the <code>reviewThreads</code> connection with the data that’s returned using that query. By using the <code>id</code> returned in <code>replyTo</code>, you should be able to find the parent comment ID that the reply was to, and build the discussion that way. </p>
<p>Taking the output of that query and plugging it into this hastily written python:</p>
<pre><code>import json
import collections
# d = Raw GraphQL query output
resp = json.loads(d, strict=False)
reviews = resp['data']['repository']['pullRequest']['reviews']['nodes']
toplevel, replies = {}, collections.defaultdict(list)

for r in reviews:
    for c in r['comments']['nodes']:
        if c['replyTo']:
            replies[c['replyTo']['id']].append(c)
        else:
            toplevel[c['id']] = c


for c in toplevel:
    print('Thread:\n')
    print(toplevel[c]['body'] + '\n')
    for r in replies[c]:
        print(' Reply: %s\n' % r['body'])
</code></pre>
<p>Using the Issue I used earlier, <a href="https://github.com/GitHawkApp/GitHawk/pull/555" rel="nofollow noopener">https://github.com/GitHawkApp/GitHawk/pull/555</a>, I get the following:</p>
<pre><code>Thread:

Probably want to extract a list of known languages to an enum + separate file? And maybe we can already find a list like that somewhere?

Then we can `return Language(rawValue: fileExtension)` (or something similar, eg. when there are multiple extensions for one language) instead of this weird `switch`. :D

Thread:

Can we remove the double space, and if it fails it should default to just returning `fileName`

  Reply: I have tidied up the code and removed the `potentialLanguage`. I was planning to use that for the 'you know what' PR as Highlight.js is not correctly automatically detected the Swift language 😅

At GitHub they are using their own Ruby library [github/linguist](https://github.com/github/linguist) can't ask the Github API to send the language?

  Reply: Leaving this reply as a test for PR review threads, demonstrating how broken they are 😞

Thread:

Not sure this is necessary, at least yet? ~Also it shouldn't **only** support Swift, even if we do keep it~

Edit// See BasThomas comment
</code></pre>
<p>As a rough cut, this seems to me to be the equivalent of the reviewThreads connection I talked about releasing earlier - three total review request threads, one of which having three comments and the other only having the top-level comment. I haven’t tested it exhaustively, though I think the approach would still work.</p>
<p>If I’ve mischaracterized anything, please let me know! I’m working on getting the <code>reviewThreads</code> through code review now, but I want to make sure it will solve your needs before releasing it! </p>
<p>Separately, please reach out to <a href="mailto:partnerengineering@github.com">partnerengineering@github.com</a> and mention this thread - we’ll be expecting you, and we can set up a chat! </p>

**Comment by nporteschaikin** [comment]
<p><a class="mention" href="/nickvanw">@nickvanw</a>:</p>
<p>Hmm. The problem for us is that the replies are not being returned, so the <code>replyTo</code> field isn’t helpful. Take this thread for example:</p>
<ul>
<li>
<p>Review comment A, by me</p>
</li>
<li>
<p>Review comment B, by someone else</p>
</li>
<li>
<p>Review comment C, by me</p>
</li>
</ul>
<p>The only comment being returned by the GraphQL API is comment A, the “parent” comment.</p>

**Comment by nickvanw** [comment]
<p><a class="mention" href="/nporteschaikin">@nporteschaikin</a>: </p>
<p>Do you have an example of a public PR where you’re experiencing this? </p>
<p>The top level reviews connection will only return the top level comments, but the comments connection inside of that seems to give me all of the information I expect. Specifically, I want to make sure you’re trying the <code>reviews</code> query that I linked to in my response, and not any of the examples that you gave in your original post. </p>
<p>I’d love to find an example where I can reproduce what you’re talking about, but I’ve been going around to a few PRs and it seems to be returning what I’d expect. </p>

**Comment by nporteschaikin** [comment]
<p>Thanks for getting back to me so quickly, <a class="mention" href="/nickvanw">@nickvanw</a>. I’m experiencing this exclusively with private repositories, so I’m going to dig into the payload further and see if I’m missing something. Will report back.</p>

**Comment by nickvanw** [comment, marked as answer]
<p>Hey <a class="mention" href="/nporteschaikin">@nporteschaikin</a>!</p>
<p>Remember that reviewThreads connection I was talking about earlier? I have good news!</p>
<p>If you use the <a href="https://developer.github.com/v4/previews/#pull-requests-preview" rel="nofollow noopener">Pull Requests Preview</a>, you should be able to run this query:</p>
<pre><code>query ($owner: String!, $name: String!, $number: Int!) {
  repository(owner: $owner, name: $name) {
    pullRequest: issueOrPullRequest(number: $number) {
      __typename
      ... on PullRequest {
        reviewThreads(first: 100) {
          edges {
            node {
              comments(first: 10) {
                nodes {
                  pullRequestReview {
                    id
                  }
                  body
                }
              }
            }
          }
        }
      }
    }
  }
}
</code></pre>
<p>This was just shipped, so the documentation should be updating soon.</p>
<p>I’d love some feedback on whether this solves your issue(s)! </p>

**Comment by nporteschaikin** [comment]
<p>Hi <a class="mention" href="/nickvanw">@nickvanw</a>! I’m going to take this for a spin. Are these objects also available via the <code>PullRequestTimelineConnection</code> connection?</p>

**Comment by nickvanw** [comment]
<p>At present moment, you can only fetch these threads by looking at this particular connection. I believe re-creating the timeline exactly as it looks on GitHub will take some munging and additional work – if something looks amiss, please let me know. </p>
<p>If any other issues persist, or something doesn’t seem right, please don’t hesitate to reach out!</p>

**Comment by iasemenov** [comment]
<p>Is there any reason why it’s</p>
<pre><code>PullRequest {
    reviewThreads
}
</code></pre>
<p>and not</p>
<pre><code>PullRequest {
    reviews {
        nodes {
            threads
        }
    }
}
</code></pre>
<p>?</p>
<p>Also it would be great to have review ID in the thread itself.</p>

**Comment by namanshergill** [comment]
<p>Hi! I don’t think <code>reviewThreads</code> serves the purpose properly when someone is trying to build a timeline from <code>timelineItems</code> until the review comment objects returned under it also have an option to include the node ID of the <code>reviewThreads</code> node it is a part of.</p>
<p>Currently, you have to paginate through all the data under <code>reviewThreads</code> and compare the top level comment ID of the review comment given in <code>timelineItems</code> through the results to be able to consistently fetch the replies to that comment. That is honestly, not really the best method, UX wise, and you end up having to wait longer for the replies to a review comment in the timeline to load the further you go into the timeline as you have to paginate through more data of <code>reviewThreads</code> until you find the comment.</p>
<p>You have to use hacky approaches like in the following:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/NamanShergill/diohub/blob/69b13ab65f5ffef13e448d4f3241ae997a3bfdc9/lib/services/pulls/pulls_service.dart#L149">
  <header class="source">

      <a href="https://github.com/NamanShergill/diohub/blob/69b13ab65f5ffef13e448d4f3241ae997a3bfdc9/lib/services/pulls/pulls_service.dart#L149" target="_blank" rel="noopener nofollow ugc">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/NamanShergill/diohub/blob/69b13ab65f5ffef13e448d4f3241ae997a3bfdc9/lib/services/pulls/pulls_service.dart#L149" target="_blank" rel="noopener nofollow ugc">NamanShergill/diohub/blob/69b13ab65f5ffef13e448d4f3241ae997a3bfdc9/lib/services/pulls/pulls_service.dart#L149</a></h4>



  <pre class="onebox">    <code class="lang-dart">
      <ol class="start lines" start="139" style="counter-reset: li-counter 138 ;">
          <li>      name: name,</li>
          <li>    ),</li>
          <li>  ),</li>
          <li>  cacheOptions: CacheManager.defaultGQLCache(refresh: refresh),</li>
          <li>);</li>
          <li>final parsed = ReviewThreadFirstCommentQuery$Query.fromJson(res.data!);</li>
          <li>
          </li>
<li>if (parsed.repository!.pullRequest!.reviewThreads.edges!.isNotEmpty) {</li>
          <li>  for (final thread</li>
          <li>      in parsed.repository!.pullRequest!.reviewThreads.edges!) {</li>
          <li class="selected">    if (thread?.node?.comments.nodes?.first?.id == commentID) {</li>
          <li>      return thread!;</li>
          <li>    }</li>
          <li>  }</li>
          <li>  return getPRReviewThreadID(commentID,</li>
          <li>      name: name,</li>
          <li>      owner: owner,</li>
          <li>      number: number,</li>
          <li>      cursor:</li>
          <li>          parsed.repository!.pullRequest!.reviewThreads.edges!.last!.cursor,</li>
          <li>      refresh: refresh);</li>
      </ol>
    </code>
  </pre>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
